### PR TITLE
Workspace symbols for JAX-RS endpoints

### LINKED ISSUE #87: Navigation code with MicroProfile (WorkspaceSymbol)
Provide the same feature than Spring Boot https://code.visualstudio.com/docs/java/java-editing#_navigating-code-with-spring-boot

### PR DESCRIPTION
**Pull request by datho7561**
Closes #87

Signed-off-by: David Thompson <davthomp@redhat.com>

### CODE REVIEW

**Comment by datho7561** [comment]
I think that Jessica's PR should be merged before mine so that I can use some of the methods she introduces or modifies in `JaxRsConstants` and `JaxRsUtils`.

**Comment by angelozerr** [comment]
@datho7561 could you add a little demo please in your PR, just to have an overview of your PR which looks exciting. Thanks!

**Comment by datho7561** [comment]
I need to do some work on this PR; it only seems to work if one of the files in the project is open.

**Comment by datho7561** [comment]
Here is a demo. I use the key combo `Ctrl+T` to open the workspace symbols view.

lsp4mp-workspace-symbols

**Comment by angelozerr** [comment]
Nice! It looks like promising!

**Comment by datho7561** [comment]
> I need to do some work on this PR; it only seems to work if one of the files in the project is open.

It looks like this is a limitation of how lsp4mp handles projects. I'll take a look into fixing it.

**Comment by datho7561** [comment]
I'm running into a `ClassCastException` when invoking the JDT search engine when I try to run the completed Quarkus super heros demo application:

```log
java.lang.ClassCastException: class org.eclipse.jdt.internal.core.JavaElementInfo cannot be cast to class org.eclipse.jdt.internal.core.ClassFileInfo (org.eclipse.jdt.internal.core.JavaElementInfo and org.eclipse.jdt.internal.core.ClassFileInfo are in unnamed module of loader org.eclipse.osgi.internal.loader.EquinoxClassLoader @6feeb453)
	at org.eclipse.jdt.internal.core.ModularClassFile.getModule(ModularClassFile.java:301)
	at org.eclipse.jdt.internal.core.PackageFragmentRoot.getSourceModuleDescription(PackageFragmentRoot.java:928)
	at org.eclipse.jdt.internal.core.PackageFragmentRoot.getModuleDescription(PackageFragmentRoot.java:904)
	at org.eclipse.jdt.internal.core.JarPackageFragmentRoot.getModuleDescription(JarPackageFragmentRoot.java:307)
	at org.eclipse.jdt.internal.core.search.matching.JavaSearchNameEnvironment.addModuleClassPathInfo(JavaSearchNameEnvironment.java:290)
	at org.eclipse.jdt.internal.core.search.matching.JavaSearchNameEnvironment.mapToClassPathLocation(JavaSearchNameEnvironment.java:284)
	at org.eclipse.jdt.internal.core.search.matching.JavaSearchNameEnvironment.computeClasspathLocations(JavaSearchNameEnvironment.java:197)
	at org.eclipse.jdt.internal.core.search.matching.JavaSearchNameEnvironment.<init>(JavaSearchNameEnvironment.java:91)
	at org.eclipse.jdt.internal.core.search.matching.IndexBasedJavaSearchEnvironment.create(IndexBasedJavaSearchEnvironment.java:27)
	at org.eclipse.jdt.internal.core.search.matching.MatchLocator.initialize(MatchLocator.java:1249)
	at org.eclipse.jdt.internal.core.search.matching.MatchLocator.locateMatches(MatchLocator.java:1287)
	at org.eclipse.jdt.internal.core.search.matching.MatchLocator.locateMatches(MatchLocator.java:1400)
	at org.eclipse.jdt.internal.core.search.matching.MatchLocator.locateMatches(MatchLocator.java:1542)
	at org.eclipse.jdt.internal.core.search.JavaSearchParticipant.locateMatches(JavaSearchParticipant.java:135)
	at org.eclipse.jdt.internal.core.search.BasicSearchEngine.findMatches(BasicSearchEngine.java:251)
	at org.eclipse.jdt.internal.core.search.BasicSearchEngine.search(BasicSearchEngine.java:602)
	at org.eclipse.jdt.core.search.SearchEngine.search(SearchEngine.java:670)
	at org.eclipse.lsp4mp.jdt.internal.jaxrs.java.JaxRsWorkspaceSymbolParticipant.collectSymbols(JaxRsWorkspaceSymbolParticipant.java:83)
	at org.eclipse.lsp4mp.jdt.internal.core.ls.MicroProfileDelegateCommandHandlerForJava.getWorkspaceSymbolsForJava(MicroProfileDelegateCommandHandlerForJava.java:486)
	at org.eclipse.lsp4mp.jdt.internal.core.ls.MicroProfileDelegateCommandHandlerForJava.executeCommand(MicroProfileDelegateCommandHandlerForJava.java:107)
	at org.eclipse.jdt.ls.core.internal.handlers.WorkspaceExecuteCommandHandler$1.run(WorkspaceExecuteCommandHandler.java:230)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:45)
	at org.eclipse.jdt.ls.core.internal.handlers.WorkspaceExecuteCommandHandler.executeCommand(WorkspaceExecuteCommandHandler.java:220)
	at org.eclipse.jdt.ls.core.internal.handlers.JDTLanguageServer.lambda$4(JDTLanguageServer.java:552)
	at org.eclipse.jdt.ls.core.internal.BaseJDTLanguageServer.lambda$0(BaseJDTLanguageServer.java:79)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture$Completion.exec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
```

Not sure if this is an upstream issue or something related to the parameters I passed to the search engine.

File: microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/core/java/symbols/IJavaWorkspaceSymbolsParticipant.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+public interface IJavaWorkspaceSymbolsParticipant {
+
+	/**
+	 * Fill in <code>symbols</code> with workspace symbols of the given project.
+	 *
+	 * @param symbols the list of symbols to add to
+	 * @param project the project to collect workspace symbols from
+	 * @param utils   the JDT utils
+	 */
+	void collectSymbols(List<SymbolInformation> symbols, IJavaProject project, IJDTUtils utils);
```
Comment (angelozerr) [inline microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/core/java/symbols/IJavaWorkspaceSymbolsParticipant.java:34]: I would provide this API:

`collectSymbols(IJavaProject project, IJDTUtils utils, List<SymbolInformation> symbols, IProgressMonitor monitor);`

I try to set 

 * read parameters as first
 * write parameters after read parameters
 * IProgressMonitor at the end.

File: microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/core/jaxrs/JaxRsUtils.java [hunk 2]
```diff
@@ -75,13 +77,17 @@ public static String getJaxRsPathValue(IAnnotatable annotatable) throws JavaMode
 	 */
 	public static String getJaxRsApplicationPathValue(IAnnotatable annotatable) throws JavaModelException {
 		IAnnotation annotationPath = getAnnotation(annotatable, JAVAX_WS_RS_APPLICATIONPATH_ANNOTATION);
+		if (annotationPath != null) {
```
Comment (angelozerr) [inline microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/core/jaxrs/JaxRsUtils.java:80]: I will change the signature of getAnnotation in  `getAnnotation(IAnnotable annotable, String name...)`

File: microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/core/ls/MicroProfileDelegateCommandHandlerForJava.java [hunk 3]
```diff
@@ -68,6 +77,9 @@ public class MicroProfileDelegateCommandHandlerForJava extends AbstractMicroProf
 	private static final String JAVA_DEFINITION_COMMAND_ID = "microprofile/java/definition";
 	private static final String JAVA_DIAGNOSTICS_COMMAND_ID = "microprofile/java/diagnostics";
 	private static final String JAVA_HOVER_COMMAND_ID = "microprofile/java/hover";
+	private static final String JAVA_WORKSPACE_SYMBOLS_ID = "microprofile/java/workspaceSymbols";
+	
+	private final IJavaWorkspaceSymbolsParticipant symbolsParticipant = new JaxRsWorkspaceSymbolParticipant();
```
Comment (angelozerr) [inline microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/core/ls/MicroProfileDelegateCommandHandlerForJava.java:82]: To be consistent with another participant, it should be better to provide an extension point.

File: microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/core/utils/ProjectUtil.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+import org.eclipse.core.resources.ResourcesPlugin;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.jdt.core.IJavaProject;
+import org.eclipse.jdt.core.JavaCore;
+
+/**
+ * A copy of some methods in {@code org.eclipse.jdt.ls.core.internal.ProjectUtils}
+ * from jdt.ls.
+ */
+public class ProjectUtil {
```
Comment (angelozerr) [inline microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/core/utils/ProjectUtil.java:28]: Please remove this class. It exists JDTMicroProfileUtils which should provide similar method and if you need, add new method in this class.

File: microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/jaxrs/java/JaxRsWorkspaceSymbolParticipant.java [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+							if (match.getElement() instanceof IAnnotation) {
+								annotatables.add((IAnnotatable) ((IAnnotation) match.getElement()).getParent());
+							} else if (match.getElement() instanceof IAnnotatable) {
+								annotatables.add((IAnnotatable) match.getElement());
+							}
+						}
+
+					}, null);
+		} catch (CoreException | ClassCastException e) {
+			MicroProfileCorePlugin.getDefault().getLogger().error(
```
[the remark this answers is not in the document]
Comment (datho7561) [reply microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/jaxrs/java/JaxRsWorkspaceSymbolParticipant.java:94]: From what I recall, using a `java.util.Logger` sends the logs somewhere where it's hard to retrieve them

**Comment by angelozerr** [comment]
@datho7561 please fix conflicts.

File: microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/core/MicroProfileCorePlugin.java [hunk 6]
```diff
@@ -57,6 +62,10 @@ public static MicroProfileCorePlugin getDefault() {
 		return plugin;
 	}
 
+	public ILog getLogger() {
+		return Platform.getLog(context.getBundle());
```
Comment (angelozerr) [inline microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/core/MicroProfileCorePlugin.java:66]: Please don't use ILog in this PR. I think it should b ebetter to do that in another PR by replacing all LOGGER with ILog. We need to dicuss more if it is better to do that.

File: microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/core/java/symbols/JavaWorkspaceSymbolsDefinition.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+	public JavaWorkspaceSymbolsDefinition(IConfigurationElement element) {
+		super(element);
+	}
+
+	@Override
+	public void collectSymbols(IJavaProject project, IJDTUtils utils, List<SymbolInformation> symbols,
+			IProgressMonitor monitor) {
+		try {
+			getParticipant().collectSymbols(project, utils, symbols, monitor);
+		} catch (CoreException e) {
```
Comment (angelozerr) [inline microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/core/java/symbols/JavaWorkspaceSymbolsDefinition.java:43]: Do like other AbstractJavaFeatureDefinition by catching Exception and not only CoreException to avoid breaking the another symbol participants process.

File: microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/core/ls/MicroProfileDelegateCommandHandlerForJava.java [hunk 8]
```diff
@@ -459,4 +463,21 @@ private static MicroProfileJavaHoverParams createMicroProfileJavaHoverParams(Lis
...
+				monitor);
+	}
+
+	private static String createMicroProfileJavaWorkspaceSymbolParams(List<Object> arguments, String commandId) {
+		Object projectUriObj = (String) arguments.get(0);
+		if (projectUriObj == null || !(projectUriObj instanceof String)) {
+			throw new UnsupportedOperationException(String
+					.format("Command '%s' must be called with one projectUri: String argument!", commandId));
+		}
+		String projectUri = (String)projectUriObj;
```
Comment (angelozerr) [inline microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/core/ls/MicroProfileDelegateCommandHandlerForJava.java:480]: `return (String)projectUriObj;`

File: microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/ls/MicroProfileWorkspaceService.java [hunk 9]
```diff
@@ -39,4 +55,36 @@ public void didChangeWatchedFiles(DidChangeWatchedFilesParams params) {
 
 	}
 
+	@Override
+	public CompletableFuture<Either<List<? extends SymbolInformation>, List<? extends WorkspaceSymbol>>> symbol(
+			WorkspaceSymbolParams params) {
+		return javaTextDocuments.getWorkspaceProjects() //
```
Comment (angelozerr) [inline microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/ls/MicroProfileWorkspaceService.java:61]: We need really having a cancelChecker here since search can be long to do and if user quit the search.

**Comment by datho7561** [comment]
I can't reproduce the test failure

File: microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/ls/MicroProfileWorkspaceService.java (hunk 9 above)
Comment (datho7561) [reply microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/ls/MicroProfileWorkspaceService.java:61]: https://github.com/eclipse/eclipse.jdt.ls/issues/2415

**Comment by datho7561** [comment]
test this please

**Comment by datho7561** [comment]
darn, looks like that's not set up on this repo

File: microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/jaxrs/java/JaxRsWorkspaceSymbolParticipant.java [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+		IMethod method = (IMethod) annotatable;
+		try {
+			if (!Flags.isPublic(method.getFlags())) {
+				return;
+			}
+		} catch (JavaModelException e) {
+			return;
+		}
+
+		String endpointUri = getEndpointUriFromAnnotatable(annotatable, applicationPrefix);
```
Comment (angelozerr) [inline microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/jaxrs/java/JaxRsWorkspaceSymbolParticipant.java:122]: I tried to play with your PR with https://github.com/FroMage/quarkus-renarde-todo/blob/main/src/main/java/rest/Todos.java and it doesn't work. 

It seems it is because methods doesn't use @Path although the method is annotated with @POST. If there are no @Path, we should use the method name if I remember.

Comment (angelozerr) [reply microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/jaxrs/java/JaxRsWorkspaceSymbolParticipant.java:122]: Ok let's see that in an another PR.

**Comment by angelozerr** [comment]
Thank so much @datho7561 for your great contribution!

File: microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/jaxrs/java/JaxRsWorkspaceSymbolParticipant.java (hunk 10 above)
Comment (datho7561) [reply microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/jaxrs/java/JaxRsWorkspaceSymbolParticipant.java:122]: I took a look at how Renard figures out which methods are REST endpoints, and it looks like any public method in a class that implements io.quarkiverse.renarde.Controller counts as a REST endpoint. Generally, the logic for calculating the method name looks different enough that I think it's worthwhile to provide a mechanism for quarkus-ls to disable the default WorkspaceSymbol support when it detects Renarde, and instead use the Renard support contributed through quarkus-ls.
